Incremental SMT2 support for enum values

esteffin opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

Commit:
Add excluded cbmc/enum regression tests

Commit:
Moved incremental smt struct encoding to subfolder

As support for non-native C types (e.g. enums) is added in the
incremental smt2 solver, other new encoding functions will be added. For
this reason, to avoid putting all the encoding functions in the
incremental smt2 solver root folder a new `encoding` folder has been
created and the struct encoding functions have been moved there.

Commit:
Regression tests for enums in incr SMT2 solver

thomasspriggs commented on src/solvers/smt2_incremental/smt2_incremental_decision_procedure.h:
```diff
@@ -13,13 +13,14 @@
 #include <memory>
 #include <unordered_map>
 
+#include "solvers/smt2_incremental/encoding/struct_encoding.h"
```
This probably needs to be `#include <solvers/smt2_incremental/encoding/struct_encoding.h>` to fix your build issues.

thomasspriggs replied:
I think the fix for this has been squashed into the wrong commit.

esteffin replied:
Yep, changed

thomasspriggs commented on src/solvers/smt2_incremental/construct_value_expr_from_smt.cpp:
```diff
@@ -70,6 +75,15 @@ class value_expr_from_smt_factoryt : public smt_term_const_downcast_visitort
       result = from_integer(bit_vector_constant.value(), type_to_construct);
       return;
     }
+    if(
```
This new feature of `construct_value_expr_from_smt.cpp` is inverse counter part of the lowering rather than part of the lowering itself. Therefore it does not match the commit title and I would prefer to see it in its own separate commit.

esteffin replied:
Done

thomasspriggs commented on src/solvers/smt2_incremental/encoding/enum_encoding.h:
```diff
@@ -0,0 +1,14 @@
+#include <util/expr.h>
+#include <util/namespace.h>
+
+/// Function to lower c_enum_tag_typet to its enum underlying type.
```
:pick: The description doesn't quite make sense as it describes a change to types, but the function takes an expression. It would be nice to have an explanation here to cover a couple of points. Firstly that the function will lower enum expressions which are sub-expressions of `expr`, rather than just `expr` itself. Secondly, what the `ns` parameter is used for. I remember right now that it is used for looking up the underlying types of enum types, but the reminder would be good.

thomasspriggs commented on src/solvers/smt2_incremental/encoding/enum_encoding.cpp:
```diff
@@ -0,0 +1,31 @@
+    work_queue.pop();
+    if(
+      const auto c_enum_type =
+        type_try_dynamic_cast<c_enum_tag_typet>(current.type()))
```
:exclamation: This doesn't look like it will work for cases where the enum type is nested inside other types. Please consider how this will/should work for examples of nested types such as an array of enum typed elements, or a struct value with an enum typed member. These should work for type tree structures which are of any arbitrary depth.

thomasspriggs replied:
Looks like this has been sorted.

thomasspriggs commented on src/solvers/smt2_incremental/encoding/enum_encoding.cpp:
```diff
@@ -0,0 +1,31 @@
+      work_queue.push(&operand);
+  }
+  return expr;
+}
```
:pick: Missing new line. Can you setup your text editor to add a trailing new line on save if it is missing from the end of the file? If I remember correctly, this doesn't just add :no_entry: signs in the github UI, it can make merges/rebases a little more awkward as well.

thomasspriggs reviewed (commented):
A few initial comments.

Commit:
Support for enum in incremental SMT2 traces

This commit enables the generation of traces containing expression of
type `c_enum_tag_typet`.

Commit:
Tests for incremental SMT2 solver enum support

thomasspriggs commented on unit/solvers/smt2_incremental/construct_value_expr_from_smt.cpp:
```diff
@@ -31,10 +31,40 @@ static mp_integer max_int(const std::size_t bits)
+    c_enum_typet::c_enum_membert member;
+    member.set_identifier("V" + std::to_string(i));
+    member.set_base_name("V" + std::to_string(i));
+    // Note: The value will be correctly set to a bv type when we know
```
Is this note actually true or is it the result of cut-and-paste? `set_value` takes an `irep_idt` / string, therefore it has formatting rather than a type.

esteffin replied:
Yes, changed.

thomasspriggs commented on unit/solvers/smt2_incremental/construct_value_expr_from_smt.cpp:
```diff
@@ -31,10 +31,40 @@ static mp_integer max_int(const std::size_t bits)
+    member.set_base_name("V" + std::to_string(i));
+    // Note: The value will be correctly set to a bv type when we know
+    // the width of the bitvector
+    member.set_value(integer2string(i));
```
:boom: `integer2string` produces a string wise representation of the given integer in base 10 by default. The value should be set using `integer2bvrep` see line 1383 of `src/ansi-c/c_typecheck_type.cpp`. The difference between these functions is currently that one uses base 10 and the other uses base 16, so there is less of a difference if your numbers are single digit only, which currently happens to be the case. Note that `integer2bvrep` should be used consistently, in case we ever change the base. If memory serves me correctly, we formerly used base 2.
:pick: It might be good to have some enum constants which have value greater than 16 in order to confirm the handling of base in all the relevant places is correct.

esteffin replied:
Agree.
Changed

thomasspriggs commented on unit/solvers/smt2_incremental/construct_value_expr_from_smt.cpp:
```diff
@@ -31,10 +31,40 @@ static mp_integer max_int(const std::size_t bits)
 
+static symbolt make_c_enum_symbol(int underlying_size)
+{
+  const signedbv_typet underlying_type{42};
```
Shouldn't this be using `underlying_size` rather than hard coded to 42? You could of course remove the parameter and make all the enums in this example the same width...

esteffin replied:
Fair point.
Changed

thomasspriggs commented on unit/solvers/smt2_incremental/construct_value_expr_from_smt.cpp:
```diff
@@ -31,10 +31,40 @@ static mp_integer max_int(const std::size_t bits)
+  symbol_tablet symbol_table;
   const namespacet ns{symbol_table};
+  const symbolt enum_symbol = make_c_enum_symbol(42);
+  const symbolt enum_tag_symbol = make_c_enum_tag_symbol(enum_symbol);
```
:pick: I think there may be better names to use here than `enum_symbol` and `enum_tag_symbol`. To me the key thing which distinguishes these is that one of them defines the type and the other one defines a value which has enum type. So maybe `enum_type_symbol` and `enum_value_symbol`.

esteffin replied:
Agree.
Changed to `enum_type_symbol` and `enum_tag_value_symbol`

thomasspriggs commented on unit/solvers/smt2_incremental/construct_value_expr_from_smt.cpp:
```diff
@@ -81,6 +111,9 @@ TEST_CASE("Value expr construction from smt.", "[core][smt2_incremental]")
         integer2bvrep(12, 64), pointer_typet(void_type(), 64 /* bits */))},
+    rowt{
+      smt_bit_vector_constant_termt{2, 42},
+      constant_exprt{"2", make_c_enum_tag_symbol(make_c_enum_symbol(42)).type}},
```
:pick: `make_c_enum_tag_symbol(make_c_enum_symbol(42)).type` is a fairly round about way of saying `enum_tag_typet{"my_enum"}` or `enum_tag_typet(enum_symbol.name)`
:no_entry_sign:  Given that the conversion now takes the namespace, we should ensure that symbols which are made are stored in the symbol table. I think this currently works due to identical symbols being constructed and added outside the `GENERATE`, but the current logic seems unnecessarily difficult to follow / maintain.
:information_source: `GENERATE` doesn't allow you to refer to variables outside of the macro invocation, because it gets transformed into a capture less lambda. If you use `GENERATE_REF` instead, then a lambda with default reference capture is used instead. This allows you to refer to symbols/types etc which are declared outside the `GENERATE_REF`.

esteffin replied:
Sensible.
Changed to `GENERATE_REF` and using `enum_tag_typet(enum_symbol.name)` instead

thomasspriggs commented on unit/solvers/smt2_incremental/construct_value_expr_from_smt.cpp:
```diff
@@ -31,10 +31,40 @@ static mp_integer max_int(const std::size_t bits)
   return power2(bits) - 1;
 }
 
+static symbolt make_c_enum_symbol(int underlying_size)
```
I suggest refactoring this function to return a `c_enum_typet` instead of a `symbol` and renaming the function as appropriate. The reason being that storing the type in the symbol upcasts it from `c_enum_typet` to `type`. It would be useful to have the original type in scope for the remainder of the test.

esteffin replied:
Yes, but the type is not really needed, so I would prefer simplicity.

thomasspriggs commented on unit/solvers/smt2_incremental/construct_value_expr_from_smt.cpp:
```diff
@@ -31,10 +31,40 @@ static mp_integer max_int(const std::size_t bits)
+  return type_symbolt{"my_enum", enum_type, ID_C};
+}
+
+static symbolt make_c_enum_tag_symbol(const symbolt &enum_symbol)
```
:bulb: Consider removing this function and inlining the code which is currently inside it. Think about how easy is it to understand what `make_c_enum_tag_symbol(enum_symbol)` does vs `symbolt{"my_enum_value", c_enum_tag_typet{enum_symbol.name}, ID_C}`? Although the former is much shorter, it requires referring to the implementation in order to understand it because the symbol name is hidden in the implementation.

esteffin replied:
Personally I prefer one well-named function instead of a "fragile" call with sub-calls.
Also it may help in the future if we decide to change this for some reason.

However I improved the name of the function.

thomasspriggs commented on unit/solvers/smt2_incremental/encoding/enum_encoding.cpp:
```diff
@@ -28,50 +29,61 @@ static c_enum_typet make_c_enum_type(const typet &underlying_type)
+  int value_count;
+  using rowt = std::pair<int, int>;
+  std::tie(size, value_count) =
+    GENERATE(rowt{42, 5}, rowt{2, 5}, rowt{0, 5}, rowt{42, 1}, rowt{42, 0});
```
:+1: I like the use of `GENERATE` here as it makes it easier to see what has and has not been tested. :thinking: This allows me to easily spot and point out that the most commonly found architectures are likely to have enums which are 64 or 32 bits wide, to match the size of int on that platform. Also, I would expect 0 bits wide to cause an error in the subsequent expression to term conversion. So it would actually be preferable if there was an INVARIANT in the lowering on the number of bits wide being greater than 0, in order to aid with fault finding by failing earlier.

esteffin replied:
Added `32` and `64` as sizes.
I will squash eventually
I removed the `0`-sized case as it should never happen before and the invariant should be checked later in the decision procedure code.

thomasspriggs commented on unit/solvers/smt2_incremental/encoding/enum_encoding.cpp:
```diff
@@ -28,50 +29,61 @@ static c_enum_typet make_c_enum_type(const typet &underlying_type)
+  std::tie(size, value_count) =
+    GENERATE(rowt{42, 5}, rowt{2, 5}, rowt{0, 5}, rowt{42, 1}, rowt{42, 0});
+  SECTION(
+    "Checking enum of underlying size " + std::to_string(size) + " with " +
```
:information_source: There is an `INFO` macro in catch which can be used to add this kind of information without increasing the indent of all the following test code.

esteffin replied:
Agree.
Changed.

thomasspriggs commented on unit/solvers/smt2_incremental/encoding/enum_encoding.cpp:
```diff
@@ -28,50 +29,61 @@ static c_enum_typet make_c_enum_type(const typet &underlying_type)
+      const array_typet lowered_array_type{
+        enum_underlying_type, from_integer(2, unsignedbv_typet{32})};
+      const array_exprt original_array_expr{
+        {from_integer(3, unsignedbv_typet{32}),
```
:question: Shouldn't the elements in this input array have enum type rather than a bit vector type?

esteffin replied:
Good catch!
Fixed

thomasspriggs commented on src/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp:
```diff
@@ -348,12 +348,15 @@ static optionalt<smt_termt> get_identifier(
+  const namespacet &ns)
 {
-  const auto handle_find_result = expression_handle_identifiers.find(expr);
+  const exprt lowered_expr = lower_enum(expr, ns);
```
I am not going to request any further changes around this issue in this PR as it seems to work as-is and I suspect its a wider issue than the enum lowering alone, but it is somewhat inconsistent that we are lowering the enums here, but not doing the other flavours of lowering. I think there are other inconsistencies around whether the expressions which are placed in the maps have been lowered or not. This is something I intend to investigate further myself.

esteffin replied:
True, but the aim of make it consistent will be done in a subsequent PR.

thomasspriggs commented on unit/solvers/smt2_incremental/encoding/enum_encoding.cpp:
```diff
@@ -28,50 +29,61 @@ static c_enum_typet make_c_enum_type(const typet &underlying_type)
 }
 
-TEST_CASE("enum encoding of expressions", "[core][smt2_incremental]")
+TEST_CASE("enum corner cases", "[core][smt2_incremental]")
```
I preferred the previous test description, as maybe we shouldn't label the entirety of the enum tests as corner cases.

esteffin replied:
Yes, restored

thomasspriggs commented:
My remaining comments are mostly qualms about the unit tests, but the code they are testing + the regression tests look good to me. The "Improved tests for enum in incremental smt2 solver" looks like an improvement and I am happy for it to be squashed into the appropriate unit test commit before merging. I'd like to get that include switch around in the first couple of commits fixed as well, so it doesn't introduce a potential issue and fix it within the same PR.

Commit:
Lower enum_tags to its enum underlying type.

In order to support expressions with type `c_enum_tag_typet` in the
incremental SMT2 solver, a new `lower_enum` has been added.
Such function lowers the type of each expression node in the AST with
type `c_enum_tag_type` to the underlying type of the `c_enum_typet` that
the tag points to.
